improver_tests/utilities/test_interpolate.py

```diff
@@ -0,0 +1,217 @@
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the functions within interpolation.py"""
+
+import unittest
+
+import numpy as np
+from iris.tests import IrisTest
+
+from improver.utilities.interpolation import interpolate_missing_data
+
+
+class Test_interpolate_missing_data(IrisTest):
+    """Test the interpolate_missing_data method"""
+    def setUp(self):
+        """ Set up arrays for testing."""
+        self.data = np.array([[1.0, 1.0, 2.0],
+                              [1.0, np.nan, 2.0],
+                              [1.0, 2.0, 2.0]])
+        self.limit_data = np.full((3, 3), 3.)
+        self.valid_data = np.full((3, 3), True)
+
+    def test_basic_linear(self):
+        """Test when all the points around the missing data are the same."""
+        data = np.ones((3, 3))
+        data[1, 1] = np.nan
+        expected = np.array([[1.0, 1.0, 1.0],
+                             [1.0, 1.0, 1.0],
+                             [1.0, 1.0, 1.0]])
+
+        data_updated = interpolate_missing_data(data)
+
+        self.assertArrayEqual(data_updated, expected)
+
+    def test_basic_nearest(self):
+        """Test when all the points around the missing data are the same."""
+        data = np.ones((3, 3))
+        data[1, 1] = np.nan
+        expected = np.array([[1.0, 1.0, 1.0],
+                             [1.0, 1.0, 1.0],
+                             [1.0, 1.0, 1.0]])
+
+        data_updated = interpolate_missing_data(data, method='nearest')
+
+        self.assertArrayEqual(data_updated, expected)
+
+    def test_different_data_for_linear_interpolation(self):
+        """Test result when linearly interpolating using points around the
+        missing data with different values."""
+        expected = np.array([[1.0, 1.0, 2.0],
+                             [1.0, 1.5, 2.0],
+                             [1.0, 2.0, 2.0]])
+
+        data_updated = interpolate_missing_data(self.data)
+
+        self.assertArrayEqual(data_updated, expected)
+
+    def test_different_data_for_nearest_neighbour(self):
+        """Test result when using nearest neighbour using points around the
+        missing data with different values."""
+        expected = np.array([[1.0, 1.0, 2.0],
+                             [1.0, 1.0, 2.0],
+                             [1.0, 2.0, 2.0]])
+
+        data_updated = interpolate_missing_data(self.data, method='nearest')
+
+        self.assertArrayEqual(data_updated, expected)
+
+    def test_too_few_points_to_linearly_interpolate(self):
+        """Test that when there are not enough points to fill the gaps using
+        linear interpolation we recover the input data. This occurs if there
+        are less than 3 points available to use for the interpolation."""
+        data = np.array([[np.nan, 1, np.nan],
+                         [np.nan, np.nan, np.nan],
+                         [np.nan, 1, np.nan]])
+
+        data_updated = interpolate_missing_data(data)
+
+        self.assertArrayEqual(data_updated, data)
```

Comment:
Might be worth defining `expected_data = data.copy()` before operation - covers for the function changing the inputs.  Also for tests below.

Reply:
I've added .copy() to the inputs so the original is guaranteed to be left unchanged. It's just a bit shorter.